scipy/stats/_qmc.py

```diff
@@ -1266,6 +1266,159 @@ def _random_oa_lhs(self, n: IntNumber = 4) -> np.ndarray:
+       :arxiv:`1706.02808`, 2017.
+    .. [3] "Low-Discrepancy sequence - Hammersley set". Wikipedia.
+       https://en.wikipedia.org/wiki/Low-discrepancy_sequence#Hammersley_set
+
+    Examples
+    --------
+    Generate samples from a low discrepancy sequence of Hammersley.
+
+    >>> from scipy.stats import qmc
+    >>> sampler = qmc.Hammersley(d=2, scramble=False)
+    >>> sample = sampler.random(n=5)
+    >>> sample
+    array([[0.   , 0.2  ],
+           [0.5  , 0.4  ],
+           [0.25 , 0.6  ],
+           [0.75 , 0.8  ],
+           [0.125, 1.   ]])
+
+    Compute the quality of the sample using the discrepancy criterion.
+
+    >>> qmc.discrepancy(sample)
+    0.0595017361111112
+
+    `random` may be called again, but because of the way the Hammersley
+    sequence is defined, the last column will still be based on linear sweep.
+
+    >>> sample_continued = sampler.random(n=5)
+    >>> sample_continued
+    array([[0.3125, 0.2   ],
+           [0.8125, 0.4   ],
+           [0.1875, 0.6   ],
+           [0.6875, 0.8   ],
+           [0.4375, 1.    ]])
+
+    Finally, samples can be scaled to bounds.
+
+    >>> l_bounds = [0, 2]
+    >>> u_bounds = [10, 5]
+    >>> qmc.scale(sample_continued, l_bounds, u_bounds)
+    array([[3.125, 2.6  ],
+           [8.125, 3.2  ],
+           [1.875, 3.8  ],
+           [6.875, 4.4  ],
+           [4.375, 5.   ]])
+
+    """
+
+    def __init__(
+        self, d: IntNumber, *, scramble: bool = True,
+        optimization: Optional[Literal["random-cd", "lloyd"]] = None,
+        seed: SeedType = None
+    ) -> None:
+        # Used in `scipy.integrate.qmc_quad`
+        self._init_quad = {'d': d, 'scramble': True,
+                           'optimization': optimization}
+        super().__init__(d=d, optimization=optimization, seed=seed)
+        self.rng = seed
+        self.scramble = scramble
+        self._halton = Halton(max(d-1, 0), scramble=scramble,
+                              seed=copy.deepcopy(seed))
+
+    def _random(
+        self, n: IntNumber = 1, *, workers: IntNumber = 1
+    ) -> np.ndarray:
+        """Draw `n` in the closed interval ``[0, 1]``.
+
+        Parameters
+        ----------
+        n : int, optional
+            Number of samples to generate in the parameter space. Default is 1.
+        workers : int, optional
+            Number of workers to use for parallel processing. If -1 is
+            given all CPU threads are used. Default is 1. It becomes faster
+            than one worker for `n` greater than :math:`10^3`.
+
+        Returns
+        -------
+        sample : array_like (n, d)
+            QMC sample.
+
+        """
+        out = np.empty((n, self.d))
+        if self.d == 0:
+            return out
+
+        out[:, :-1] = self._halton.random(n, workers=workers)
+        out[:, -1] = np.arange(1, n+1) / n
+        if self.scramble:
+            out[:, -1] -= self.rng.random(size=n) / n
```

Comment:
LHS like, not considering other dimensions. I think it makes sense 👍

Reply:
I'm just following the definitions I've found.

https://en.wikipedia.org/wiki/Low-discrepancy_sequence#Hammersley_set

https://idaes-pse.readthedocs.io/en/1.8.0/user_guide/modeling_extensions/surrogate/pysmo/pysmo_hammersley.html

https://people.sc.fsu.edu/~jburkardt/f_src/hammersley/hammersley.html

The thing they seem to disagree on is exactly what the linear sweep is and which column it should be in.

![image](https://user-images.githubusercontent.com/6570539/190841242-075c59d6-a200-4598-8fdb-84d08e91f28a.png)

![image](https://user-images.githubusercontent.com/6570539/190841252-295699d8-fb94-45b1-bad0-6daf2e6cc24b.png)

![image](https://user-images.githubusercontent.com/6570539/190841285-3a47c2cd-685e-4fc1-b672-a5f56f9b2248.png)

Reply:
I think this is the best description:
https://www.pbr-book.org/3ed-2018/Sampling_and_Reconstruction/The_Halton_Sampler#HammersleyandHaltonSequences

Because now I actually understand the Halton sequence, too. Anyway, I changed it so that the linear sweep is the first column; now I think we match this reference exactly.

Reply:
Yes the choice of the column is one of the reasons I did not yet try to code this myself. What you had did make sense to me as the last dimension is less good as the previous ones. And in an application it does not matter much as you can map dimensions to your labels as you wish.